### DISCUSSION TITLE
Rust desk ALT+TAB problem

### DISCUSSION THREAD

**Discussion by mbrojo** [rustdesk/rustdesk#10022] [Q&A]
On the client i use several application at once.
Often on the client. I need to to use ALT+TAB to switch between previous and current one.

When I work on the client without Rust Desk connection, it works perfect.
The PROBLEM is that when I am connected through rust desk to this client, the ALT+TAB is not working as desired, but it switches through all the open windows not only between the last two ones.

What am i doing wrong?

**Comment by fufesou** [comment]
> The PROBLEM is that when I am connected through rust desk to this client
> switches through all the open windows not only between the last two ones.

There're no differences for me. How does it look like? Is it possible to give a short video?

What's OS of the controlled side?

**Reply by mbrojo** [reply]
I don't know how exatly I could record videos.
To make it more clear - when I am connected via RustDesk and press ALT+TAB, it behaves as if I pressed ALT+ESC instead.

**Reply by mbrojo** [reply]
I mean I know how to record video, but I don't know how I could record a video which would show what shortcut I pressed..

**Reply by fufesou** [reply]
> it behaves as if I pressed ALT+ESC instead

Windows? 

> The PROBLEM is that when I am connected through rust desk to this client, the ALT+TAB is not working as desired

`ALT` + `TAB` is pressed on the controlling side or just the controlled side?

**Reply by mbrojo** [reply]
> `ALT` + `TAB` is pressed on the controlling side or just the controlled side?

I believe it's pressed on the controlled side.

**Reply by mbrojo** [reply]
> Windows?

Yes - both computers are Windows 10

**Reply by fufesou** [reply]
> I believe it's pressed on the controlled side.

I still cannot reproduce.

How about connect and do not press any keys on the controlling side?
I mean, is it still affects the controlled side, if no keys are pressed(on the controling side) in that connection.

**Reply by mbrojo** [reply]
> > I believe it's pressed on the controlled side.
> 
> I still cannot reproduce.
> 
> How about connect and do not press any keys on the controlling side? I mean, is it still affects the controlled side, if no keys are pressed(on the controling side) in that connection.

Maybe I get you inaccurate information. I press ALT+TAB on the keyboard on controlling side. I expect this shortcut to transmit to controlled side. But on the controlled side it behaves as if ALT+ESC was pressed.

**Comment by fufesou** [comment]
Could you run the attached keyboard grab app on the controlled side?

Please run it before a new connection is established. The press `ALT` + `TAB`. You will see what keys are pressed and released.

Please attach the screenshot here.

![Screenshot 2024-11-23 162523](https://github.com/user-attachments/assets/24bf7dbc-326c-4021-87c1-24508c3a964d)


[grab.zip](https://github.com/user-attachments/files/17879030/grab.zip)

**Reply by mbrojo** [reply]
> Could you run the attached keyboard grab app on the controlled side?

I don't want to be rude, but I don't want to run .exe. Could you send me the source code?

**Reply by fufesou** [reply]
Ok

https://github.com/rustdesk-org/rdev/blob/master/examples/grab.rs

**Reply by fufesou** [reply]
https://keyboard-test.space/

This site is also good to test. Though it can not record all events of `Alt` + `Tab`.

**Reply by mbrojo** [reply]
so i ran the grab, here is the result:

![image](https://github.com/user-attachments/assets/2350ed7a-0480-4715-853c-9b0c5f2786b6)

**Reply by fufesou** [reply]
> so i ran the grab, here is the result:

Do you press `Alt` + `Tab` on the controlling side?

**Reply by mbrojo** [reply]
> > so i ran the grab, here is the result:
> 
> Do you press `Alt` + `Tab` on the controlling side?

Just to be sure: I am sitting on the controlling side (computer1). I am connected to the controlled side (computer2).
Since I am sitting on the controlling side the only way to press the key combination is to press it on the physical keyboard at the controlling side. So for me yes - I press Alt + Tab on the controlling side.

Am I missing something?

**Reply by fufesou** [reply]
So it stange that only `Alt` and `Tab` are simulated. I have no idea at this time. 🥲 

> Do you press `Alt` + `Tab` on the controlling side?

I ask it because there's always one more `KeyRelease` event for me.

**Reply by mbrojo** [reply]
> So it stange that only `Alt` and `Tab` are simulated. I have no idea at this time. 🥲
> 
> > Do you press `Alt` + `Tab` on the controlling side?
> 
> I ask it because there's always one more `KeyRelease` event for me.

I don't get it

**Reply by fufesou** [reply]
![image](https://github.com/user-attachments/assets/874f15d6-ced5-420a-ab08-e7471322bff8)

There're two release events of `Tab`

**Reply by mbrojo** [reply]
> There're two release events of `Tab`

I see. I checked this also on Controlling side. And I only get one release events of Tab also there

**Reply by mbrojo** [reply]
Any update on this? I tried the grab.exe on another computer and also there I get the KeyReleaste(Tab) only once.

**Reply by fufesou** [reply]
Sorry, but to be honest, I have no clue.

Because the screentshot shows that only `Alt` and `Tab` are pressed and released.

> KeyReleaste(Tab) only once

This does not matter actually. I asked it because I though it should be twice if RustDesk triggers the events.

**Reply by mbrojo** [reply]
I finally fixed it.

On the Controlled side, there was this setting in Registry: 
AltTabSettings = 1 in the key HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer
This caused the Alt+tab work in 'oldway', which means not showing windows preview when alttab was pressed and besides this alt+tab stopped working properly via RustDesk. 
When I switched it to 0, the new way was activated and alt+tab via RustDesk works again.

Thanks for your effort a lot.

**Reply by hervehobbes** [reply]
Hello, I have the same problem when using Ubuntu Desktop 26.04 and connecting to Windows 11.
When connected to Windows and using Alt Tab, I'm switching to Linux Desktop.
I'm using keyboard as much as possible since 40 years and it's "annoying"

**Reply by fufesou** [reply]
@hervehobbes What version do you use on Ubuntu? https://github.com/rustdesk/rustdesk/pull/14302 should have fixed it. Please try the latest version.

**Comment by hervehobbes** [comment]
Thanks for your answer.
Linux : Version: 1.4.1
Date de compilation: 2025-07-29 04:21
Empreinte numérique: 5408 dea5 9e50 019d c9b5 6dd3 f6a5 8da1 fb21 f995 cdee 1d06 9ffa 27a9 58b5 4b45

Windows:
1.4.8.29700619
